clippy_lints/src/excessive_bools.rs

```diff
@@ -0,0 +1,196 @@
+
+fn is_bool_ty(ty: &Ty) -> bool {
+    if let TyKind::Path(None, path) = &ty.kind {
+        return match_path_ast(path, &["bool"]);
+    }
+    false
+}
+
+fn check_fn(cx: &EarlyContext<'_>, fn_sig: &FnSig, span: Span, max_fn_params_bools: u64) {
+    match fn_sig.header.ext {
+        Extern::Implicit | Extern::Explicit(_) => return,
+        Extern::None => (),
+    }
+
+    if max_fn_params_bools
+        < fn_sig
+            .decl
+            .inputs
+            .iter()
+            .filter(|param| is_bool_ty(&param.ty))
+            .count()
+            .try_into()
+            .unwrap()
+    {
+        span_lint_and_help(
+            cx,
+            FN_PARAMS_EXCESSIVE_BOOLS,
+            span,
+            "excessive amount of bools in function parameters",
+            "consider refactoring bools into two-variant enums",
+        );
+    }
+}
+
+impl EarlyLintPass for ExcessiveBools {
+    fn check_item(&mut self, cx: &EarlyContext<'_>, item: &Item) {
+        if in_macro(item.span) {
+            return;
+        }
+        match &item.kind {
+            ItemKind::Struct(variant_data, _) => {
+                if attr_by_name(&item.attrs, "repr").is_some() {
+                    return;
+                }
+
+                if self.max_struct_bools
+                    < variant_data
+                        .fields()
+                        .iter()
+                        .filter(|field| is_bool_ty(&field.ty))
+                        .count()
+                        .try_into()
+                        .unwrap()
+                {
+                    span_lint_and_help(
+                        cx,
+                        STRUCT_EXCESSIVE_BOOLS,
+                        item.span,
+                        "excessive amount of bools in a struct",
+                        "consider using a state machine or refactoring bools into two-variant enums",
+                    );
+                }
+            },
+            ItemKind::Impl {
+                of_trait: None, items, ..
+            }
+            | ItemKind::Trait(_, _, _, _, items) => {
+                let mut visitor = AssocFnVisitor {
+                    cx,
+                    max_fn_params_bools: self.max_fn_params_bools,
+                };
+                for item in items {
+                    visitor.visit_assoc_item(item);
+                }
+            },
+            ItemKind::Fn(fn_sig, _, _) => check_fn(cx, fn_sig, item.span, self.max_fn_params_bools),
+            _ => (),
+        }
+    }
+}
+
+struct AssocFnVisitor<'a> {
+    cx: &'a EarlyContext<'a>,
+    max_fn_params_bools: u64,
+}
+
+impl<'a> Visitor<'a> for AssocFnVisitor<'a> {
+    fn visit_assoc_item(&mut self, assoc_item: &AssocItem) {
+        if let AssocItemKind::Fn(fn_sig, _) = &assoc_item.kind {
+            check_fn(self.cx, fn_sig, assoc_item.span, self.max_fn_params_bools);
+        }
+
+        // dont go down into children because `check_item` is going to cover them
```

Comment:
In that case a visitor is overkill, since you only check items in the first layer, which you could also do in the for loop above.